Source/Dafny/Util.cs

```diff
@@ -465,4 +467,666 @@ class IEnumerableComparer<T> : IEqualityComparer<IEnumerable<T>> {
+        return UsesSpecFeatures(e.E0) || UsesSpecFeatures(e.E1) || UsesSpecFeatures(e.E2);
+      } else if (expr is LetExpr) {
+        var e = (LetExpr)expr;
+        if (e.Exact) {
+          MakeGhostAsNeeded(e.LHSs);
+          return UsesSpecFeatures(e.Body);
+        } else {
+          Contract.Assert(e.RHSs.Count == 1);
+          if (UsesSpecFeatures(e.RHSs[0])) {
+            foreach (var bv in e.BoundVars) {
+              bv.MakeGhost();
+            }
+          }
+          return UsesSpecFeatures(e.Body);
+        }
+      } else if (expr is QuantifierExpr) {
+        var e = (QuantifierExpr)expr;
+        Contract.Assert(e.SplitQuantifier == null); // No split quantifiers during resolution
+        return e.UncompilableBoundVars().Count != 0 || UsesSpecFeatures(e.LogicalBody());
+      } else if (expr is SetComprehension) {
+        var e = (SetComprehension)expr;
+        return !e.Finite || e.UncompilableBoundVars().Count != 0 || (e.Range != null && UsesSpecFeatures(e.Range)) || (e.Term != null && UsesSpecFeatures(e.Term));
+      } else if (expr is MapComprehension) {
+        var e = (MapComprehension)expr;
+        return !e.Finite || e.UncompilableBoundVars().Count != 0 || UsesSpecFeatures(e.Range) || (e.TermLeft != null && UsesSpecFeatures(e.TermLeft)) || UsesSpecFeatures(e.Term);
+      } else if (expr is LambdaExpr) {
+        var e = (LambdaExpr)expr;
+        return UsesSpecFeatures(e.Term);
+      } else if (expr is WildcardExpr) {
+        return false;
+      } else if (expr is StmtExpr) {
+        var e = (StmtExpr)expr;
+        return UsesSpecFeatures(e.E);
+      } else if (expr is ITEExpr) {
+        ITEExpr e = (ITEExpr)expr;
+        return UsesSpecFeatures(e.Test) || UsesSpecFeatures(e.Thn) || UsesSpecFeatures(e.Els);
+      } else if (expr is NestedMatchExpr) {
+        return UsesSpecFeatures(((NestedMatchExpr)expr).ResolvedExpression);
+      } else if (expr is MatchExpr) {
+        MatchExpr me = (MatchExpr)expr;
+        if (UsesSpecFeatures(me.Source)) {
+          return true;
+        }
+        return me.Cases.Exists(mc => UsesSpecFeatures(mc.Body));
+      } else if (expr is ConcreteSyntaxExpression) {
+        var e = (ConcreteSyntaxExpression)expr;
+        return e.ResolvedExpression != null && UsesSpecFeatures(e.ResolvedExpression);
+      } else if (expr is SeqConstructionExpr) {
+        var e = (SeqConstructionExpr)expr;
+        return UsesSpecFeatures(e.N) || UsesSpecFeatures(e.Initializer);
+      } else if (expr is MultiSetFormingExpr) {
+        var e = (MultiSetFormingExpr)expr;
+        return UsesSpecFeatures(e.E);
+      } else {
+        Contract.Assert(false); throw new cce.UnreachableException();  // unexpected expression
+      }
+    }
+    static void MakeGhostAsNeeded(List<CasePattern<BoundVar>> lhss) {
+      foreach (CasePattern<BoundVar> lhs in lhss) {
+        MakeGhostAsNeeded(lhs);
+      }
+    }
+
+    static void MakeGhostAsNeeded(CasePattern<BoundVar> lhs) {
```

Comment:
Is this called during "pass 2" of the Resolver? If so, it should be mentioned in that nice list you added, too.

Reply:
These functions are called by `UsesSpecFeatures` which is called in various places (including the `CheckIsCompilable` and I'm don't know what it is doing or why.